File: nvidia/c-cpp/cusparse/examples/sparse_cpu.c
#include <mkl_spblas.h>

#include <math.h>
#include <stdio.h>
#include <stdlib.h>

static int make_poisson2d_csr(int nx, int ny, int *row, int *col, double *val) {
  int p = 0, offset = 0;
  for (int iy = 0; iy < ny; ++iy)
    for (int ix = 0; ix < nx; ++ix) {
      p = iy * nx + ix;
      row[p] = offset;
      if (iy > 0) {
        col[offset] = p - nx;
        val[offset++] = -1.0;
      }
      if (ix > 0) {
        col[offset] = p - 1;
        val[offset++] = -1.0;
      }
      col[offset] = p;
      val[offset++] = 4.0;
      if (ix + 1 < nx) {
        col[offset] = p + 1;
        val[offset++] = -1.0;
      }
      if (iy + 1 < ny) {
        col[offset] = p + nx;
        val[offset++] = -1.0;
      }
    }
  row[nx * ny] = offset;
  return offset;
}

int main(void) {
  const int nx = 1024, ny = 1024, n = nx * ny, nnz = 5 * n - 2 * nx - 2 * ny;
  int *row = (int *)malloc((size_t)(n + 1) * sizeof(*row));
  int *col = (int *)malloc((size_t)nnz * sizeof(*col));
  double *val = (double *)malloc((size_t)nnz * sizeof(*val));
  double *x = (double *)malloc((size_t)n * sizeof(*x));
  double *y = (double *)malloc((size_t)n * sizeof(*y));
  sparse_matrix_t matrix = NULL;
  int status = EXIT_FAILURE;
  double max_error = 0.0;
  if (!row || !col || !val || !x || !y) {
    fprintf(stderr, "CSR allocation failed\n");
    goto cleanup;
  }
  if (make_poisson2d_csr(nx, ny, row, col, val) != nnz) {
    fprintf(stderr, "CSR generation failed\n");
    goto cleanup;
  }
  for (int i = 0; i < n; ++i) {
    x[i] = 1.0;
    y[i] = 1.0;
  }
  struct matrix_descr descr;
  descr.type = SPARSE_MATRIX_TYPE_GENERAL;
  descr.mode = SPARSE_FILL_MODE_FULL;
  descr.diag = SPARSE_DIAG_NON_UNIT;
  if (mkl_sparse_d_create_csr(&matrix, SPARSE_INDEX_BASE_ZERO, n, n, row,
                              row + 1, col, val) != SPARSE_STATUS_SUCCESS ||
      mkl_sparse_d_mv(SPARSE_OPERATION_NON_TRANSPOSE, 1.0, matrix, descr, x,
                      1.0, y) != SPARSE_STATUS_SUCCESS) {
    fprintf(stderr, "oneMKL Sparse SpMV failed\n");
    goto cleanup;
  }
  for (int iy = 0; iy < ny; ++iy)
    for (int ix = 0; ix < nx; ++ix) {
      int neighbors =
          4 - (ix == 0) - (ix + 1 == nx) - (iy == 0) - (iy + 1 == ny);
      double expected = 5.0 - neighbors;
      max_error = fmax(max_error, fabs(y[iy * nx + ix] - expected));
    }
  printf("CPU Poisson CSR SpMV complete; max error = %.17g\n", max_error);
  status = max_error <= 1.0e-12 ? EXIT_SUCCESS : EXIT_FAILURE;

cleanup:
  if (matrix != NULL &&
      mkl_sparse_destroy(matrix) != SPARSE_STATUS_SUCCESS) {
    fprintf(stderr, "mkl_sparse_destroy failed\n");
    status = EXIT_FAILURE;
  }
  free(y);
  free(x);
  free(val);
  free(col);
  free(row);
  return status;
}

File: nvidia/fortran/cusparse/examples/sparse_cpu.f90
include 'mkl_spblas.f90'

program sparse_cpu
    use mkl_spblas
    use, intrinsic :: ieee_arithmetic, only: ieee_is_finite
    implicit none
    integer :: allocation_status

    interface
        subroutine make_poisson2d_csr( &
            nx, ny, nrow, ncol, nnz, &
            row_offsets, col_indices, values)

            implicit none

            integer, intent(in) :: nx, ny
            integer, intent(out) :: nrow, ncol, nnz

            integer, allocatable, intent(out) :: &
                row_offsets(:), col_indices(:)
            real(8), allocatable, intent(out) :: values(:)
        end subroutine make_poisson2d_csr
    end interface

    integer, parameter :: nx = 1024
    integer, parameter :: ny = 1024
    real(8), parameter :: alpha = 1.0d0
    real(8), parameter :: beta  = 1.0d0
    integer :: nrow, ncol, nnz
    integer :: istat
    integer, allocatable :: row_offsets(:), col_indices(:)
    real(8), allocatable :: values(:), vec_x(:), vec_y(:)
    type(sparse_matrix_t) :: mat_A
    type(matrix_descr) :: descr

    ! 1始まりのCSR配列を生成する
    call make_poisson2d_csr( &
        nx, ny, nrow, ncol, nnz, &
        row_offsets, col_indices, values)

    allocate(vec_x(ncol), vec_y(nrow), stat=allocation_status)

    if (allocation_status /= 0) error stop 'allocation failed'
    vec_x = 1.0d0; vec_y = 1.0d0

    descr%type = SPARSE_MATRIX_TYPE_GENERAL
    descr%mode = SPARSE_FILL_MODE_FULL
    descr%diag = SPARSE_DIAG_NON_UNIT

    istat = mkl_sparse_d_create_csr( &
        mat_A, SPARSE_INDEX_BASE_ONE, &
        nrow, ncol, &
        row_offsets(1:nrow), &
        row_offsets(2:nrow + 1), &
        col_indices, values)
    if (istat /= 0) error stop 'mkl_sparse_d_create_csr failed'

    istat = mkl_sparse_d_mv( &
        SPARSE_OPERATION_NON_TRANSPOSE, &
        alpha, mat_A, descr, &
        vec_x, beta, vec_y)
    if (istat /= 0) error stop 'mkl_sparse_d_mv failed'

    block
        integer :: ix, iy, p, neighbors
        real(8) :: expected
        if (.not. all(ieee_is_finite(vec_y))) error stop 'nonfinite SpMV result'
        do iy = 1, ny
            do ix = 1, nx
                p = (iy-1)*nx+ix
                neighbors = 0
                if (ix > 1) neighbors = neighbors+1
                if (ix < nx) neighbors = neighbors+1
                if (iy > 1) neighbors = neighbors+1
                if (iy < ny) neighbors = neighbors+1
                expected = alpha*real(4-neighbors,8)+beta
                if (abs(vec_y(p)-expected) > 1.0d-10) error stop 'SpMV verification failed'
            end do
        end do
    end block
    print *, 'verification PASS'
    istat = mkl_sparse_destroy(mat_A)
    if (istat /= 0) error stop 'mkl_sparse_destroy failed'

    deallocate(row_offsets, col_indices, values)
    deallocate(vec_x, vec_y)
end program sparse_cpu
